Add Py::get for GIL-independent access to frozen classes. (#3158)
PyO3/pyo3: guide/src/class.md

});
```

### frozen classes: Opting out of interior mutability

> As detailed abvoe, runtime borrow checking is currently enabled by default. But a class can opt of out it by declaring itself `frozen`. It can still use interior mutability via standard Rust types like `RefCell` or `Mutex`, but it is not bound to the implementation provided by PyO3 and can choose the most appropriate strategy on field-by-field basis.

Reviewer: ```suggestion
As detailed above, runtime borrow checking is currently enabled by default. But a class can opt of out it by declaring itself `frozen`. It can still use interior mutability via standard Rust types like `RefCell` or `Mutex`, but it is not bound to the implementation provided by PyO3 and can choose the most appropriate strategy on field-by-field basis.
```

> Classes which are `frozen` and also `Sync`, e.g. they do use `Mutex` but not `RefCell`, can also be accessed independently of the acquiring the GIL via the `PyCell::get` and `Py::get` methods:

Reviewer: ```suggestion
Classes which are `frozen` and also `Sync`, e.g. they do use `Mutex` but not `RefCell`, can be accessed without needing the Python GIL via the `PyCell::get` and `Py::get` methods:
```

```rust
use std::sync::atomic::{AtomicUsize, Ordering};
# use pyo3::prelude::*;

#[pyclass(frozen)]
struct FrozenCounter {
    value: AtomicUsize,
}

> let cell  = Python::with_gil(|py| {
>     let counter = FrozenCounter { value: AtomicUsize::new(0) };
> 
>     Py::new(py, counter).unwrap()
> });
> 
> cell.get().value.fetch_add(1, Ordering::Relaxed);

Reviewer: It might be helpful to add a type annotation here to show the user what's going on:

```suggestion
let py_frozen_counter: Py<FrozenCounter> = Python::with_gil(|py| {
    let counter = FrozenCounter { value: AtomicUsize::new(0) };

    Py::new(py, counter).unwrap()
});

py_frozen_counter.get().value.fetch_add(1, Ordering::Relaxed);
```

```

> In the long run, frozen classes are poised to become the new default thereby guiding the PyO3 ecosystem towards a more deliberate application of interior mutability. Eventually, this should enable further optimizations of PyO3's internals and avoid downstream code paying the cost of interior mutability even though it is not actually required.

Reviewer: I think it's fine to have this though I may suggest not committing to a timeline:

```suggestion
Frozen classes are likely to become the default thereby guiding the PyO3 ecosystem towards a more deliberate application of interior mutability. Eventually, this should enable further optimizations of PyO3's internals and avoid downstream code paying the cost of interior mutability when it is not actually required.
```
